include/kernels/coarse_op_kernel.cuh

```diff
@@ -120,74 +128,176 @@ namespace quda {
       : Y(Y), X(X), Y_atomic(Y_atomic), X_atomic(X_atomic),
       UV(UV), AV(AV), U(U), V(V), C(C), Cinv(Cinv), spin_bs(spin_bs_), spin_map(),
       kappa(static_cast<Float>(kappa)), mass(static_cast<Float>(mass)), mu(static_cast<Float>(mu)), mu_factor(static_cast<Float>(mu_factor)),
       fineVolumeCB(V.VolumeCB()), coarseVolumeCB(X.VolumeCB()),
       fine_to_coarse(fine_to_coarse), coarse_to_fine(coarse_to_fine),
-      bidirectional(bidirectional), shared_atomic(false), parity_flip(shared_atomic ? true : false),
-      aggregates_per_block(1), max_d(nullptr)
+      bidirectional(bidirectional), shared_atomic(false), parity_flip(false),
+        aggregates_per_block(1), max_h(nullptr), max_d(nullptr), max(nullptr)
     {
       if (V.GammaBasis() != QUDA_DEGRAND_ROSSI_GAMMA_BASIS)
         errorQuda("Gamma basis %d not supported", V.GammaBasis());
 
       for (int i=0; i<QUDA_MAX_DIM; i++) {
         x_size[i] = x_size_[i];
         xc_size[i] = xc_size_[i];
-        geo_bs[i] = geo_bs_[i];
+        geo_bs[i] = x_size[i] / xc_size[i];
         comm_dim[i] = comm_dim_partitioned(i);
       }
     }
   };
 
+  /** This is a wrapper class that derives from the kernel argument
+      struct, and is used to enable the maximum element computation */
+  template <typename T>  struct ArgMax : public T {
+    static constexpr bool compute_max = true;
+    ArgMax(const T& t) : T(t) { }
+  };
+
   /**
-     Calculates the matrix UV^{s,c'}_mu(x) = \sum_c U^{c}_mu(x) * V^{s,c}_mu(x+mu)
-     Where: mu = dir, s = fine spin, c' = coarse color, c = fine color
-     or, if dir == QUDA_IN_PLACE, UV^{s,c'}(x) = \sum_c C^{c}_mu(x) * V^{s,c}_mu(x+mu)
+     @brief Helper for computing if the present site is within the
+     forwards halo region
+     @param[in] coord Grid coordinates
+     @param[in] dim Dimension of the shift
+     @param[in] nFace Depth of the halo
+     @param[in] arg Kernel argument
   */
-  template<int dim, QudaDirection dir, typename Wtype, typename Arg>
-  __device__ __host__ inline void computeUV(Arg &arg, const Wtype &Wacc, int parity, int x_cb, int i0, int j0)
+  template <typename Arg> constexpr bool isHalo(const int coord[], int dim, int nFace, const Arg &arg)
   {
-    int coord[4];
-    getCoords(coord, x_cb, arg.x_size, parity);
+    switch (dim) {
+    case 0: return arg.comm_dim[0] && coord[0] + nFace >= arg.x_size[0];
+    case 1: return arg.comm_dim[1] && coord[1] + nFace >= arg.x_size[1];
+    case 2: return arg.comm_dim[2] && coord[2] + nFace >= arg.x_size[2];
+    case 3: return arg.comm_dim[3] && coord[3] + nFace >= arg.x_size[3];
+    }
+    return false;
+  }
+
+  /**
+     @brief Helper for computing checkerboard index from from a shift
+     of length nFace in the direction dim
+     @param[in] x Grid coordinates
+     @param[in] X Grid dimensions
+     @param[in] dim Dimension of the shift
+     @param[in] nFace Depth of the halo
+  */
+  template <typename I, typename Coord>
```

Comment:
Move to `index_helper.cuh`

Reply:
c94109308c46ca051eea11a865a2f445d990e257